mono/metadata/class.c

```diff
@@ -1889,6 +1889,10 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 
 		if (!(field->type->attrs & FIELD_ATTRIBUTE_STATIC)) {
 			ftype = mono_type_get_underlying_type (field->type);
+			if (!ftype) {
+				mono_class_set_type_load_failure (klass, "Field '%s' is an enum type with a bad underlying type", field->name);
+				continue;
+			}
```

Comment:
Well... for each static field we only check it once (`class.c:2194` when we compute `static_field_refs`).

For non-static fields, we check once here and once in the various branches of the layout switch where I assume we want to successfully layout the class fields even if we ultimately will mark the class as failed.  I could get rid of the `mono_class_set_type_load_failure()` here, I suppose.